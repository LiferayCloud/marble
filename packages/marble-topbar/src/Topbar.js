'use strict';

import Component from 'metal-component';
import Soy from 'metal-soy';
import Toggler from 'metal-toggler';
import {Config} from 'metal-state';

import templates from './Topbar.soy.js';

/**
 * Topbar component.
 */
class Topbar extends Component {
  attached() {
    this.toggler = new Toggler({
      content: '.topbar-list',
      header: '.topbar-toggle',
      expandedClasses: 'topbar-list-expanded'
    });
  }

  disposed() {
    let toggler = this.toggler;

    if (toggler) {
      toggler.dispose();
    }
  }
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
Topbar.STATE = {
  /**
   * Additional CSS classes to be added
   * @type {!String}
   * @default undefined
   */
  style: Config.string(),

  /**
   * Defines how the logo should look like
   * @type {?Object|undefined}
   * @default undefined
   */
  logo: Config.shapeOf({
    href: Config.string(),
    icon: Config.string(),
    image: Config.string(),
    text: Config.string(),
  }),

  /**
   * The list of menu items
   * @type {?Array|undefined}
   * @default undefined
   */
  items: Config.arrayOf(
    Config.shapeOf({
      href: Config.string(),
      label: Config.string(),
      selected: Config.bool(),
      variant: Config.string(),
    })
  ).value([]),
};

Soy.register(Topbar, templates);

export {Topbar};
export default Topbar;
