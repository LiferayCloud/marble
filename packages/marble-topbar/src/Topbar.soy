{namespace Topbar}

/**
 * @param items
 * @param? logo
 * @param? style
 */
{template .render}
  {let $styleClasses: $style ? 'topbar ' + $style : 'topbar' /}

  <nav class="{$styleClasses}">
    {call .logo}
      {param logo: $logo /}
    {/call}

    {call .menu}
      {param items: $items /}
    {/call}
  </nav>
{/template}

/**
 * @param? logo
 */
{template .logo}
  <div class="topbar-logo">
    {let $href: $logo and $logo.href ? $logo.href : '/' /}

    <a class="topbar-logo-link" href="{$href}">
      {if $logo and $logo.icon}
        <span class="topbar-logo-icon {$logo.icon}"></span>
      {/if}

      {if $logo and $logo.image}
        <img class="topbar-logo-image" src="{$logo.image}">
      {/if}

      {if $logo and $logo.text}
        <h3 class="topbar-logo-text">{$logo.text}</h3>
      {/if}
    </a>
  </div>
{/template}

/**
 * @param items
 */
{template .menu}
  {if length($items)}
    <nav class="topbar-menu">
      <button class="topbar-toggle">Menu</button>
      <ul class="topbar-list">
        {foreach $item in $items}
          {let $localVariant: $item.variant ?: 'default' /}

          {delcall Topbar.item variant="$localVariant"}
            {param href: $item.href /}
            {param label: $item.label /}
            {param target: $item.target /}
            {param type: $item.type /}
            {param selected: $item.selected /}
          {/delcall}
        {/foreach}
      </ul>
    </nav>
  {/if}
{/template}

/**
 * @param href
 * @param label
 * @param target
 * @param type
 * @param? selected
 */
{deltemplate Topbar.item variant="'default'"}
  {let $selectedClass: $selected ? 'topbar-link-selected' : '' /}

  <li class="topbar-item">
    {if $type == 'button'}
      <a class="btn btn-accent" href="{$href}" target="{$target}">
        {$label}
      </a>
    {else}
      <a class="topbar-link {$selectedClass}" href="{$href}" target="{$target}">
        <span>{$label}</span>
      </a>
    {/if}
  </li>
{/deltemplate}
