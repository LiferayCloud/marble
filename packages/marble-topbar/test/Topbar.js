import Topbar from '../src/Topbar';

let topbar;

describe('Topbar', () => {
  afterEach(() => {
    if (topbar) {
      topbar.dispose();
    }
  });

  it('should generate the default markup', () => {
    topbar = new Topbar();

    expect(topbar).toMatchSnapshot();
  });

  it('should render with an additional class', () => {
    topbar = new Topbar({
      style: 'topbar-light'
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with href', () => {
    topbar = new Topbar({
      logo: {
        href: '#'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with text', () => {
    topbar = new Topbar({
      logo: {
        text: 'Marble'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with icon', () => {
    topbar = new Topbar({
      logo: {
        icon: 'icon-12-github'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with icon and text', () => {
    topbar = new Topbar({
      logo: {
        text: 'Marble',
        icon: 'icon-12-github'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with image', () => {
    topbar = new Topbar({
      logo: {
        image: 'logo.svg'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a logo with image and text', () => {
    topbar = new Topbar({
      logo: {
        text: 'Marble',
        image: 'logo.svg'
      }
    });

    expect(topbar).toMatchSnapshot();
  });

  it('should render a list of navigation items', () => {
    topbar = new Topbar({
      items: [
        {
          label: 'Apps',
          href: '/apps',
          selected: true
        },
        {
          label: 'Blog',
          href: '/blog',
          target: '_blank'
        },
        {
          label: 'Login',
          href: '/login',
          type: 'button'
        }
      ]
    });

    expect(topbar).toMatchSnapshot();
  });
});
