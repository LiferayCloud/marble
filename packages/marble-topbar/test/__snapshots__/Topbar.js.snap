// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topbar should generate the default markup 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/"></a>
  </div>
</nav>
`;

exports[`Topbar should render a list of navigation items 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/"></a>
  </div>
  <nav class="topbar-menu">
    <button class="topbar-toggle">Menu</button>
    <ul class="topbar-list">
      <li class="topbar-item">
        <a class="topbar-link topbar-link-selected" href="/apps">
          <span>Apps</span>
        </a>
      </li>
      <li class="topbar-item">
        <a class="topbar-link " href="/blog">
          <span>Blog</span>
        </a>
      </li>
    </ul>
  </nav>
</nav>
`;

exports[`Topbar should render a logo with href 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="#"></a>
  </div>
</nav>
`;

exports[`Topbar should render a logo with icon 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/">
      <span class="topbar-logo-icon icon-12-github"></span>
    </a>
  </div>
</nav>
`;

exports[`Topbar should render a logo with icon and text 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/">
      <span class="topbar-logo-icon icon-12-github"></span>
      <h3 class="topbar-logo-text">Marble</h3>
    </a>
  </div>
</nav>
`;

exports[`Topbar should render a logo with image 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/">
      <img class="topbar-logo-image" src="logo.svg">
    </a>
  </div>
</nav>
`;

exports[`Topbar should render a logo with image and text 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/">
      <img class="topbar-logo-image" src="logo.svg">
      <h3 class="topbar-logo-text">Marble</h3>
    </a>
  </div>
</nav>
`;

exports[`Topbar should render a logo with text 1`] = `
<nav class="topbar">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/">
      <h3 class="topbar-logo-text">Marble</h3>
    </a>
  </div>
</nav>
`;

exports[`Topbar should render with an additional class 1`] = `
<nav class="topbar topbar-light">
  <div class="topbar-logo">
    <a class="topbar-logo-link" href="/"></a>
  </div>
</nav>
`;
